### ISSUE TITLE
Figure out how to ensure fairness between clients

### ISSUE THREAD

**Issue by howardjohn** [istio/ztunnel#24]

**Comment by daixiang0** [comment]
Is it about LB?

**Comment by stevenctl** [comment]
This involves looking at our concurrency model and configuration and also requires an evaluation method.

**Comment by howardjohn** [comment]
Tokio has some 'budget' concept but it is per-task. We would want some concept of task-groups.

But we also probably want higher level fairness like bytes/s 

https://docs.rs/tokio/latest/src/tokio/runtime/coop.rs.html

**Comment by daixiang0** [comment]
The title is confused, does it mean that Tokio is unfair in some case or we need different priorities for different clients?

**Comment by daixiang0** [comment]
> Tokio fairness is per task. ideally ours would be per-pod

Sounds like the fairness wrt bandwidth for per-pod, which is better to be done by the OS facilities. I know `tc` or some NICs can do this.

**Comment by howardjohn** [comment]
Yeah good point, I guess now that we have "in pod" mode, its much easier to do this at the CNI layer

**Comment by howardjohn** [comment]
There is still some fairness like number of connections open/establish (sending a few bytes of TLS handshake is way more expensive than a few bytes of data) that would be nice to have

**Comment by daixiang0** [comment]
Do you mean fairness in connection pool?
